add google-style and refactor

TheRoyalTnetennba opened the pull request:
* Add [google-style](https://github.com/googleapis/ruby-style) linter rules with the following exceptions:
  - Lint/ShadowingOuterLocalVariable has been disabled as methods that use Google::Gax::Operation both return and yield the same operation.
  - Lint/UselessAssignment has been disabled as assignment is sometimes done just to clarify what a method returns
  - Metrics/AbcSize has been disabled, mainly to limit the scope of work. It is unlikely that all samples would ever conform to a useful limit.
  - Metrics/BlockLength has been disabled, see Metrics/AbcSize
  - Metrics/CyclomaticComplexity has been disabled, see Metrics/AbcSize
  - Metrics/LineLength  has been disabled, see Metrics/AbcSize
  - Metrics/MethodLength  has been disabled, see Metrics/AbcSize
  - Style/BlockDelimiters has been updated with EnforcedStyle: braces_for_chaining. I'd like to include this in the google-style gem.
  - Style/EmptyMethod has been updated with nforcedStyle: expanded. I'd like to include this in the google-style gem.
  - Style/GuardClause has been disabled to improve clarity in samples. It would be useful to have a discussion about this for google-style.
  - Style/IfUnlessModifier see Style/GuardClause
  - Style/LineEndConcatenation has been disabled to improve clarity.
  - Style/MethodCallWithArgsParentheses has been updated with AllowParenthesesInChaining: true. Will likely include this in the google-style gem.
  - Style/MethodDefParentheses has been updated with EnforcedStyle: require_no_parentheses_except_multiline. It would be useful to have a discussion about this for google-style.
  - Style/Next has been disabled to improve clarity.
  - Style/SafeNavigation has been disabled to improve clarity.
  - Style/SymbolArray has been updated with EnforcedStyle: brackets. It would be useful to have a discussion about this for google-style.
  - Style/WordArray - same as Style/SymbolArray
* Refactor
* Set kokoro build scripts to use `bundle update` instead of `bundle install`

Currently, rubocop cannot actively enforce the following desired styles:
  - Style/GuardClause, Style/IfUnlessModifier, Style/Next: would like for contributors to use:
```ruby
if thing
  do_other_thing
end
```
style conditionals where reasonable.
  - Style/LineEndConcatenation: would like to require contributors to use `+`
  - Style/SafeNavigation: would like to require contributors to avoid using the safe navigation operator

frankyn commented:
This is amazing @TheRoyalTnetennba.

Responding to questions:

> %w[string string] vs ["string", "string"]

I prefer `["string", "string"]` given it's easier to parse in a sample.

> the Style/GuardClause cop

Could you provide more background here?

> Metrics/LineLength - how does docsite handle long lines, what are its limits, etc.
 
When samples are long a user has to scroll horizontally. I think the sample column length is 80, but I can't recall where we document this. 

> Add Gemfile.lock to .gitignore

We still used this to help keep a consistent test given Gemfile's don't specify a specific version. We'd need to use pessimistic versioning.

beccasaurus commented:
Swinging by with some +1's

> %w[string string] vs ["string", "string"]

+1 to @frankyn

This includes `%i[ for symbols ]` as well, I prefer `[:foo, :bar]` for Google's code samples

> Add Gemfile.lock to .gitignore

+1 to @frankyn

That said, I would be happy to be convinced otherwise ~ what's the case for ignoring lock files?

It's pretty important for us to be able to reproduce working samples. If the reason is to ensure our samples continue to work with latest (sans lock), I would argue the way to fix this is to make sure we `rm Gemfile.lock; bundle install` in CI scripts. If it starts failing, at least the debugger can get themselves back to a working state first. tl;dr there are other approaches which don't require ignoring the lock file.

beccasaurus commented on appengine/rails-cloudsql-mysql/spec/controllers/cats_controller_spec.rb:
```diff
@@ -52,81 +51,81 @@
-        }.to change(Cat, :count).by(1)
+        expect do
+          post :create, params: { cat: valid_attributes }, session: valid_session
+        end.to change(Cat, :count).by(1)
```
Oh no, please no.

`end.x` is sacrilege

~> @dazuma

TheRoyalTnetennba commented:
* Reverted to `["string", "string]` `[:symbol, :symbol]` notation
* Enforced braces for chaining blocks
* Reverted Gemfile.lock

This is probably out of scope for this PR, but is there a way we are keeping track of what samples list dependency versions and which don't? How would we know if a sample no longer works if a user naively installs latest gem version b/c the sample doesn't specify?

frankyn commented:
I don't think we have a dependency tracker for samples that list dependencies or not. At least I've never seen one.

TheRoyalTnetennba commented:
@frankyn regarding Style/GuardClause currently we are using the following rules:


> ````ruby
> # bad
> def test
>   if something
>     work
>   end
> end
> 
> # good
> def test
>   return unless something
>   work
> end
> 
> # also good
> def test
>   work if something
> end
> 
> # bad
> if something
>   raise 'exception'
> else
>   ok
> end
> 
> # good
> raise 'exception' if something
> ok
Personally, I would prefer to read this. I felt it was worth asking if y'all thought that in, the context of samples, it might be better to more consistently write in a
```ruby
if something
  work
end
```
style

frankyn commented:
Thanks @TheRoyalTnetennba,

I'd prefer having in samples:
```ruby 
if something
  work
end
```

dazuma commented:
Burn `Style/GuardClause` with fire.

dazuma commented:
Regarding `Gemfile.lock`, normally I advocate `.gitignore`ing them in most cases except for production-destined "final" products. This is because we want to be able to test against the latest dependencies, to ensure we find out about broken dependency updates. (Otherwise our users will find those before we will). But @beccasaurus 's argument about giving debuggers a known good starting point, is well taken. I'd just say that it then becomes very important that we `bundle update` in our CI scripts.

beccasaurus commented:
> But @beccasaurus 's argument about giving debuggers a known good starting point, is well taken. I'd just say that it then becomes very important that we bundle update in our CI scripts.

If we're not bundle updating in CI scripts, or might not, then I'm super +1 adding them to .gitignore. It's a blunt but very functional correction. So maybe I was wrong. Simply not having the .lock file is a very effective strategy. Except ~ yeah, it's really nice to be able to get back to a working state.

> We still used this to help keep a consistent test given Gemfile's don't specify a specific version. We'd need to use pessimistic versioning.

We have been warned. I would hate to see us, as a result, make our version numbers super pessimistic in our Gemfiles.

If the goal is to test that new users have a working experience, versions should only be specified when there's an explicit reason to, right?

Rebecca's...

Advice for production systems:

 - Use Gemfile.lock and be very mindful of versions at all times! While also bumping consistently because security!

Advice for developers:

 - Think about locking your dependencies to major versions (minor for 0.x), or else know that they _will_ one day explode.

Advice for samples:

 - Fuck it! Versions are for the weak. Bring on the pain! ` ` 📍 You are here

TheRoyalTnetennba commented:
@tbpg could spanner api be enabled?

frankyn commented:
@TheRoyalTnetennba I enabled the Spanner API for cloud-samples-ruby-test-2.

TheRoyalTnetennba commented:
Anyone have access to the 2.5 logs? I can't see why this is failing

TheRoyalTnetennba commented:
Looks like we're all green. Could I get a review? No rush, it's a big PR.

frankyn commented on CONTRIBUTING.md:
```diff
@@ -47,7 +47,8 @@ For instructions regarding development environment setup, please visit [the docu
 1. Ensure that your code adheres to the existing style in the sample to which
-   you are contributing.
+   you are contributing and that running `bundle exec rubocop` from the root
+   directory passes.
```
@TheRoyalTnetennba is there a way to format code as well when using rubocop?

frankyn commented on iot/iot.rb:
```diff
@@ -387,15 +387,15 @@
 
   puts "Devices:"
-  if response.devices && response.devices.any?
+  if response.devices&.any?
```
Is the following `response.devices&.any?`  common convention now? I find the expanded version easier to parse.

TheRoyalTnetennba replied:
I think the convention is fairly common. Both default rubocop and standardrb will fail if you don't use the safe navigation operator.

dazuma replied:
It's considered idiomatic now—although it is newer syntax, and yes it is harder to parse, especially for the old-timers like me. I think it's fine to keep it, especially since it is the rubocop default. (I'm not sold on standard, but it's probably just because I'm wary about there being 15.)

bleything replied:
Oof. Fellow old-timer here: I do not care for that syntax.

I also feel pretty strongly that code in docs and samples should be very slow to adopt new language features. These things are adopted slowly, and it takes some people a very long time to learn about them. The fact that we're talking about whether it's common 3.5 years after its release is evidence of this.

Given the choice between an older style (`response.devices && response.devices.any?`) and a newer style (`response.devices&.any?`) I think we should err on the side of the older style until the newer style is ubiquitous.

beccasaurus replied:
I enjoy `response.devices&.any?` in my codez.
I prefer `response.devices && response.devices.any?` on the website for readability.

frankyn commented on vision/product_search/product_search_add_product_to_product_set.rb:
```diff
@@ -15,12 +15,12 @@
 require "google/cloud/vision"
 
-def product_search_add_product_to_product_set(
+def product_search_add_product_to_product_set \
```
I found this change to be a little confusing when compared to other samples. I understand the usage, but I would find using parenthesis here easier to parse. (opinion).

bleything replied:
hard agree

bleything replied:
whoops, didn't realize I had an outdated section open, plz to disregard

frankyn commented on appengine/metadata_server/app.rb:
```diff
@@ -18,14 +18,14 @@
 get "/" do
   uri = URI.parse(
-    "http://metadata.google.internal/computeMetadata/v1" +
+    "http://metadata.google.internal/computeMetadata/v1" \
```
I find it easier to understand `+` than `\`, but I'm open to either just wanted to surface it as an opinion.

beccasaurus replied:
I also much prefer `+`, particularly for website samples.

Using `\` is a fine convention for application/production code, if folks really care about concatenations occurring. But I would like to + `+` for website samples.

dazuma commented on appengine/cloud-tasks/create_task.rb:
```diff
@@ -32,15 +31,13 @@ def create_task project_id, location_id, queue_id, payload: nil, seconds: nil
-  if payload
-    task[:app_engine_http_request][:body] = payload
-  end
+  task[:app_engine_http_request][:body] = payload if payload
```
Did we say we were going to drop the guard clauses?

TheRoyalTnetennba replied:
We did. Didn't catch all the ones I'd already changed. Thanks for pointing this out.

TheRoyalTnetennba replied:
BRB there's like over 100 of them :sob:

TheRoyalTnetennba replied:
There's another cop, Style/IfUnlessModifier, which covers this. Kinda broken as we aren't currently setting line limits. Gonna disable it.

TheRoyalTnetennba replied:
Style/Next also affects this

dazuma commented on appengine/datastore/app.rb:
```diff
@@ -28,7 +28,7 @@
 
   # Query the last 10 visits from the Datastore
-  query     = datastore.query("Visit").order("timestamp", :desc).limit(10)
+  query     = datastore.query("Visit").order("timestamp", :desc).limit 10
```
Is rubocop really insisting on removing those parens?

TheRoyalTnetennba replied:
Yup

bleything replied:
ah, `Style/MethodCallWithArgsParentheses`, my old nemesis. Enabling `AllowParenthesesInChaining` will permit this specific case, which I think is preferable. Mixing parenthesis style on the same line is kinda gross.

dazuma commented on appengine/rails-cloudsql-mysql/app/controllers/cats_controller.rb:
```diff
@@ -23,26 +23,24 @@ def index
   # GET /cats/1.json
-  def show
-  end
+  def show; end
```
Which cop is this? I'm not a fan...

TheRoyalTnetennba commented:
Rubocop is currently unable to enforce our rules for the following:
1. Style/GuardClause
1. Style/IfUnlessModifier
1. Style/LineEndConcatenation
1. Style/Next

bleything commented on texttospeech/synthesize_text.rb:
```diff
@@ -79,12 +79,16 @@ def synthesize_ssml ssml:
+
+
+"asdfasdfasdfasdfasdfasdfasdfasdfasdfasdasdfasdfasdfasdfasdfasdf" +
+  "asdfasdfasdfasdfasdfasdfasdfasdfasdfasdasdfasdfasdfasdfasdfasdf"
```
what's this for?

TheRoyalTnetennba replied:
Lol, trying to find the cop that determines multi-line string concatenation

bleything commented:
> Rubocop is currently unable to enforce our rules for the following:

What are our rules for those? I'm not seeing much in the style guide but maybe I'm looking at the wrong one or I'm just not processing it correctly?

I enabled each in turn and looked at the results:

* `Style/GuardClause` - no failures
* `Style/IfUnlessModifier` - some of these look legit
* `Style/LineEndConcatenation` - some of these look legit too, in particular the bigquery and spanner samples where the queries could be heredocs
* `Style/Next` - one failure that may be a subtle violation of the style guide that I'm just missing:

[dialogflow/context_management.rb:34](https://github.com/GoogleCloudPlatform/ruby-docs-samples/blob/master/dialogflow/context_management.rb#L34-L41)
```ruby
    if context.parameters
      puts "Fields:"
      context.parameters.fields.each do |field, value|
        if value.string_value
          puts "\t#{field}: #{value.string_value}"
        end
      end
    end
```

It wants that to be `next unless context.parameters`, which seems reasonable to me.

bleything commented:
It's not really relevant to this PR per se but I noticed a lot of keyword args as I was reading through the samples. Are keyword args a thing that people actually use out in the world? Basically I'm wondering all the same things as we talked about [in this thread](https://github.com/GoogleCloudPlatform/ruby-docs-samples/pull/408#discussion_r267422906).

TheRoyalTnetennba commented:
@bleything by "rules", I just meant, the desired style based on the discussion in this PR. These are all cops w/out the appropriate EnforcedStyle options - we can only disable/enable them.

Style/GuardClause - it seemed to me that consensus was to disallow if possible. If this is possible, I don't know how. I've currently settled for disabling it.

Style/IfUnlessModifier - love to hear your thoughts on it. Personally, I like it. Others didn't seem to, and the broader trend of opinions seem to be anti-code-golfy? :woman_shrugging: not that it's the code-golfiest of conventions... Anyway, the mention of lack of enforcement options stems from only being able to turn the cop off or on, no EnforcedStyle: spread_that_conditional_out option.

Style/LineEndConcatenation - no enforced style options. No use_pluses_instead_of_backslashes as others mentioned, and no make_it_a_heredoc.

Style/Next - again, love to hear your thoughts. I would prefer to see this enabled. I don't know that it would be a good idea to require people to favor conditionals over `next`, but if that's what we wanted, we can't using any of the currently provided EnforcedStyle options. 

Incidentally, there doesn't seem to be a way to require people to avoid the safe navigation operator. You can only disable Style/SafeNavigation

TheRoyalTnetennba commented:
I'd like to explore enabling Metrics/LineLength at some point, along with any other Metrics cops that make sense. I do want this to happen in a different PR. The diff is big enough as is and LineLength isn't really the same kind of discussion as other style issues.

bleything commented:
@TheRoyalTnetennba: thanks for the explanation, that all makes a lot of sense. I didn't realize how late I was to this PR and how much I'd missed, I'm sorry for the added noise.

TheRoyalTnetennba commented:
@bleything no worries. This matters to all of us and I'm happy for the feedback

TheRoyalTnetennba commented:
Anyone willing to give me the much-coveted LGTM

dazuma commented:
@frankyn You beat me by 20 seconds!!

frankyn commented:
lol!!! :smile: @dazuma
